splice_spark2/src/main/spark2.2/com/splicemachine/spark2/splicemachine/SplicemachineContext.scala

```diff
@@ -489,68 +552,269 @@ class SplicemachineContext(options: Map[String, String]) extends Serializable {
+            val lastVR = lastMsg.get.asInstanceOf[KafkaReadFunction.Message].vr
+            val hashCols = if( lastVR.length > 0) { Range(0,lastVR.length-1).toArray } else { Array(0) }
+            val lastKHash = lastVR.hashCode(hashCols)
+            def hash: Row => Int = row => externalizable(row, schema, partition).hashCode(hashCols)
+            
+            //val itr34 = itr12._2.duplicate
+            val itr34 = itrRow.duplicate   //itr12.duplicate
+            
+            val inKafka_NotInKafka = itr34._1.span( hash(_) != lastKHash )
+            if( inKafka_NotInKafka._2.hasNext ) {
+              inKafka_NotInKafka._2.next  // matches the last item in Kafka, get past it
+              //println(s"$id SMC.sendData Retry skip ${hash(r)} $lastKHash")
+              msgCount = inKafka_NotInKafka._1.size + 1  // this message count was lost during previous task failure
+              inKafka_NotInKafka._2
+            } else {  // itrRow starts after the last item added to Kafka
+              itr34._2
+            }
           }
-          msgIdx = entriesInKafka.asInstanceOf[Int]
+        } else {
+          itrRow
         }
 
         val props = new Properties
         props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, kafkaServers)
         props.put(ProducerConfig.CLIENT_ID_CONFIG, "spark-producer-s2s-smc-"+java.util.UUID.randomUUID() )
         props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, classOf[IntegerSerializer].getName)
-        props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, classOf[ExternalizableSerializer].getName)
-
-        val producer = new KafkaProducer[Integer, Externalizable](props)
-
-        while( itrRow.hasNext ) {
-          producer.send( new ProducerRecord(topicName, msgIdx, externalizable(itrRow.next, schema)) )
-          msgIdx += 1
+        props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, classOf[ByteArraySerializer].getName)
+//        // Throughput performance?
+//        println(s"SMC.sendData batch 1MB linger 750ms")
+//        props.put(ProducerConfig.BATCH_SIZE_CONFIG, (1000*1000).toString )
+//        props.put(ProducerConfig.LINGER_MS_CONFIG, "500")
+
+        val producer = new KafkaProducer[Integer, Array[Byte]](props)
+        
+        val rowK = new RowForKafka(topicName, partition, schema)
+        rowK.sparkRow = if( itr.hasNext ) {
+          msgCount += 1
+          Some(itr.next)
+        } else None
+        rowK.msgCount = msgCount
+        
+        val kryo = new KryoSerialization()
+        kryo.init
+        
+        while( itr.hasNext ) {
+          rowK.valueRow = Some(externalizable(rowK.sparkRow.get, schema, partition))
+          rowK.send(producer, kryo)
+          msgCount += 1
+          rowK.sparkRow = Some(itr.next)
+          rowK.msgCount = msgCount
         }
+        lastRowsToSend.add(rowK)
+        
+        kryo.close
 
+        insAccum.add(msgCount)
+
+        println(s"$id SMC.sendData t $topicName p $partition records $msgCount")
```

Comment:
We shouldn't print to console, use Debugging output instead

Reply:
Changed all printlns to logs in a temporary branch for review while regression tests are running, and will push to this branch after regression is finished.

https://github.com/splicemachine/spliceengine/blob/0263ed86b9c614882b6cde88a47c3caf87ca7027/splice_spark2/src/main/spark2.4/com/splicemachine/spark2/splicemachine/SplicemachineContext.scala#L799-L801

Also included some other code improvements.

Reply:
Pushed code to this branch.